File: resources/xcode/NativePHP/NativeRender/NativeRootStackRenderer.swift
import SwiftUI

/// Native chrome renderer for the `native_root_stack` sentinel. Uses
/// SwiftUI's `NavigationStack` with a `NavigationCoordinator` that
/// bridges the runtime path between PHP's router and SwiftUI's path
/// binding.
///
/// **Path sync model.** PHP's router is the source of truth. Every
/// publish at a `native_root_stack` root flows through `coordinator.receive(uri:rootNode:)`,
/// which decides push / pop / no-op based on whether the URI is already
/// in the path. The path-bound `NavigationStack` then animates the
/// transition. User-initiated swipe-pops are caught by `.onChange` of
/// `coordinator.path` and fire `sendSystemBackEvent` back to PHP so the
/// runloop pops its own stack to match.
///
/// **Per-URI cache.** Each level of the stack is rendered from the
/// coordinator's `rootNodeCache` keyed on URI. This is what makes
/// edge-swipe-back work — during the system pop animation, both the
/// from- and to-screens need real content; the cache provides it.
///
/// **Three-tier appearance** (matches the locked-in contract):
///   - No `background_color` set → defaults; iOS 26+ applies Liquid Glass
///   - `background_color` set → opaque native bar with developer's color
///   - inline `<native:top-bar>` blade bypasses native chrome entirely
struct NativeRootStackRenderer: View {
    let node: NativeUINode

    @ObservedObject private var coordinator = NavigationCoordinator.shared

    var body: some View {
        let currentUri = node.props.getString("current_uri", default: "")

        // Write cache synchronously so destinations always render from
        // the freshest tree on this very render pass. The path mutation
        // (push / pop) stays deferred via async since `path` is
        // @Published and can't be mutated during body.
        if !currentUri.isEmpty {
            coordinator.cache(uri: currentUri, node: node)
            DispatchQueue.main.async {
                coordinator.receive(uri: currentUri, rootNode: node)
            }
        }

        return NavigationStack(path: $coordinator.path) {
            // The `NavigationStack` root is the bottommost stack URI
            // (`coordinator.rootUri`) — pushed levels live in `path` and
            // are rendered via `.navigationDestination(for:)`. On the
            // very first publish, before the coordinator has seeded its
            // root, fall back to `currentUri`.
            let rootUri = coordinator.rootUri ?? currentUri
            destination(uri: rootUri, isRoot: true)
                .navigationDestination(for: String.self) { uri in
                    destination(uri: uri, isRoot: false)
                }
        }
        .onChange(of: coordinator.path) { newPath in
            coordinator.onPathChange(newPath: newPath)
        }
    }

    /// Resolve a URI to its renderable content from the cache. Cache is
    /// kept fresh by the synchronous `coordinator.cache(...)` write at
    /// the top of `body`, so this never reads a stale tree. Reading from
    /// cache (rather than from the live `node`) also keeps mid-animation
    /// destinations stable: when PHP republishes during a swipe-back,
    /// `currentUri` shifts but the popping destination keeps rendering
    /// from its own cache entry until the animation finishes.
    @ViewBuilder
    private func destination(uri: String, isRoot: Bool) -> some View {
        if let cached = coordinator.rootNodeCache[uri] {
            renderRoot(cached, isRoot: isRoot)
        } else {
            Color.clear
        }
    }

    /// Render the screen-content child of a `native_root_stack` plus the
    /// toolbar / title configured on it. Each push level has its own
    /// independent toolbar config (read from its own cached root).
    @ViewBuilder
    private func renderRoot(_ root: NativeUINode, isRoot: Bool) -> some View {
        let title = root.props.getString("title", default: "")
        let subtitle = root.props.getString("subtitle", default: "")
        let showBack = root.props.getBool("back")
        let bgArgb = root.props.getColor("background_color", default: 0)
        let textArgb = root.props.getColor("text_color", default: 0)
        let displayModeStr = root.props.getString("display_mode", default: "inline")
        let actions = root.children.filter { $0.type == "top_bar_action" }
        // Bottom-pinned content (chat input, search bar, etc.) — extracted
        // out of children so it doesn't render inline; pinned via
        // `.safeAreaInset(.bottom)` below so the keyboard pushes it up.
        let bottomBar = root.children.first { $0.type == "bottom_bar" }
        let screenContent = root.children.first {
            $0.type != "top_bar_action" && $0.type != "bottom_bar"
        }

        let textColor: Color = textArgb != 0 ? Color(argb: textArgb) : .primary
        let hasExplicitBg = bgArgb != 0

        // Map the PHP-side string to SwiftUI's NavigationBarItem.TitleDisplayMode.
        //   `large`     — iOS-native big title, left-aligned, collapses on scroll
        //   `automatic` — iOS picks (large at root, inline after a push)
        //   else        — small centered title (previous default)
        let titleDisplayMode: NavigationBarItem.TitleDisplayMode = {
            switch displayModeStr {
            case "large":     return .large
            case "automatic": return .automatic
            default:          return .inline
            }
        }()

        screenView(screenContent)
            .navigationTitle(title)
            .navigationBarTitleDisplayMode(titleDisplayMode)
            // iOS 18+ has a first-class `.navigationSubtitle(...)` that sits
            // with the title (next to it for inline, under the large title
            // for large). Use it when the OS supports it AND we're not
            // already rendering subtitle via the principal toolbar item
            // (which only happens for inline mode below).
            .modifier(NavigationSubtitleModifier(
                subtitle: subtitle,
                showsAsPrincipal: titleDisplayMode == .inline
            ))
            .toolbar {
                // Manual back chevron only at the root level — pushed
                // levels get the system back chevron from NavigationStack
                // itself, which fires the path binding shrink (caught by
                // onChange and forwarded to PHP). Showing both would
                // duplicate the chevron at every pushed level.
                if showBack && isRoot {
                    ToolbarItem(placement: .topBarLeading) {
                        Button {
                            NativeElementBridge.sendSystemBackEvent()
                        } label: {
                            Image(systemName: "chevron.backward")
                                .font(.system(size: 17, weight: .semibold))
                                .foregroundColor(textColor)
                        }
                    }
                }
                // Render subtitle as a `.principal` toolbar item ONLY when
                // displayMode is inline. With `.large` (or `.automatic` at
                // root), the principal slot duplicates content next to the
                // big title — the user sees two stacked titles. iOS 18+
                // exposes `.navigationSubtitle(...)` which sits with the
                // large title naturally; until we adopt that path, we just
                // suppress the principal subtitle for non-inline modes.
                if !subtitle.isEmpty && titleDisplayMode == .inline {
                    ToolbarItem(placement: .principal) {
                        VStack(spacing: 0) {
                            Text(title)
                                .font(.headline)
                                .foregroundColor(textColor)
                            Text(subtitle)
                                .font(.caption)
                                .foregroundColor(textColor.opacity(0.7))
                        }
                    }
                }
                ToolbarItemGroup(placement: .topBarTrailing) {
                    ForEach(actions) { action in
                        actionView(action, textColor: textColor)
                    }
                }
            }
            .modifier(StackBarBackgroundModifier(argb: bgArgb))
            .modifier(StackBottomBarInsetModifier(bottomBar: bottomBar))
    }

    @ViewBuilder
    private func screenView(_ node: NativeUINode?) -> some View {
        if let node = node {
            // GlassEffectContainer coordinates `.interactive(true)` press
            // animations across glass surfaces in this screen so they
            // crossfade between idle and pressed states cleanly. Without
            // a container, the per-glass-effect animation isn't scoped
            // and the press transition renders as a visible flicker
            // behind the touched element. iOS 26+ only.
            NodeView(node: node).withGlassContainer()
        } else {
            Color.clear
        }
    }

    /// Renders one trailing action — plain Button when the action has
    /// no sub-items, SwiftUI `Menu` of sub-items when `NavAction.items()`
    /// was set on the PHP side (which puts sub-actions in
    /// `action.children`).
    @ViewBuilder
    private func actionView(_ action: NativeUINode, textColor: Color) -> some View {
        let icon = action.props.getString("icon", default: "ellipsis")
        let subItems = action.children.filter { $0.type == "top_bar_action" }

        if subItems.isEmpty {
            Button {
                if action.onPress != 0 {
                    NativeElementBridge.sendPressEvent(action.onPress, nodeId: action.id)
                }
            } label: {
                Image(systemName: getIconForName(icon))
                    .font(.system(size: 17, weight: .semibold))
                    .foregroundColor(textColor)
            }
        } else {
            Menu {
                ForEach(subItems) { item in
                    let itemLabel = item.props.getString("label", default: "")
                    let itemIcon = item.props.getString("icon", default: "")
                    let isDestructive = item.props.getBool("destructive")
                    Button(role: isDestructive ? .destructive : nil) {
                        if item.onPress != 0 {
                            NativeElementBridge.sendPressEvent(item.onPress, nodeId: item.id)
                        }
                    } label: {
                        if !itemIcon.isEmpty {
                            Label(itemLabel, systemImage: getIconForName(itemIcon))
                        } else {
                            Text(itemLabel)
                        }
                    }
                }
            } label: {
                Image(systemName: getIconForName(icon))
                    .font(.system(size: 17, weight: .semibold))
                    .foregroundColor(textColor)
            }
        }
    }
}

/// Pins a `bottom_bar` element above the safe-area-bottom (and above
/// the keyboard when one is presented). Renders nothing if the layout
/// didn't supply a bottom bar for this level. Mirrors the tabs
/// renderer's `BottomBarInsetModifier`: iOS 26 uses `.safeAreaBar`
/// (first-class floating glass bar primitive), pre-26 falls back to
/// `.safeAreaInset(.bottom)`.
private struct StackBottomBarInsetModifier: ViewModifier {
    let bottomBar: NativeUINode?

    func body(content: Content) -> some View {
        if let bottomBar, let inner = bottomBar.children.first {
            if #available(iOS 26.0, *) {
                content.safeAreaBar(edge: .bottom) {
                    NodeView(node: inner)
                }
            } else {
                content.safeAreaInset(edge: .bottom, spacing: 0) {
                    NodeView(node: inner)
                }
            }
        } else {
            content
        }
    }
}

/// Same conditional-bar-background pattern as the tabs renderer: skip
/// `.toolbarBackground` entirely when the layout didn't supply an
/// explicit color, so iOS 26 keeps its adaptive Liquid Glass material
/// on the navigation bar instead of having `.clear` forcibly applied.
private struct StackBarBackgroundModifier: ViewModifier {
    let argb: Int

    func body(content: Content) -> some View {
        if argb != 0 {
            content
                .toolbarBackground(Color(argb: argb), for: .navigationBar)
                .toolbarBackground(.visible, for: .navigationBar)
        } else {
            content
        }
    }
}

/// Conditionally applies iOS 26+ `.navigationSubtitle(...)` so the
/// subtitle sits with the title in the system bar — the right place for
/// it when the title display mode is `.large`. Skipped when the inline
/// path already renders the subtitle via a `.principal` `ToolbarItem`
/// (so we don't double-render it).
///
/// `.navigationSubtitle` was added in iOS 26 (alongside the toolbar
/// title-display-mode work). Pre-iOS-26 the subtitle is silently dropped
/// in `.large`/`.automatic` modes — fall back to `displayMode('inline')`
/// to keep the subtitle visible on older OSes.
private struct NavigationSubtitleModifier: ViewModifier {
    let subtitle: String
    let showsAsPrincipal: Bool

    func body(content: Content) -> some View {
        if subtitle.isEmpty || showsAsPrincipal {
            content
        } else if #available(iOS 26.0, *) {
            content.navigationSubtitle(subtitle)
        } else {
            content
        }
    }
}

File: resources/androidstudio/app/src/main/java/com/nativephp/mobile/ui/nativerender/NativeRootStackRenderer.kt
package com.nativephp.mobile.ui.nativerender

import androidx.activity.compose.BackHandler
import androidx.compose.animation.AnimatedContent
import androidx.compose.animation.core.tween
import androidx.compose.animation.slideInHorizontally
import androidx.compose.animation.slideOutHorizontally
import androidx.compose.animation.togetherWith
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.snapshots.SnapshotStateList
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import com.nativephp.mobile.ui.MaterialIcon

/**
 * Compose port of iOS's `NativeRootStackRenderer`. Renders the
 * `native_root_stack` sentinel via Material 3 `Scaffold` + `TopAppBar`
 * with a path-driven AnimatedContent for push / pop transitions.
 *
 * `BackHandler` intercepts system back / predictive-back gesture,
 * shrinking the path and firing `sendSystemBackEvent` so the PHP
 * runloop pops to match.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun NativeRootStackRenderer(node: NativeUINode, modifier: Modifier = Modifier) {
    val currentUri = node.props.getString("current_uri", "")
    val coordinator = NavigationCoordinator

    // Sync-write the cache so destinations always render the freshest
    // tree on this very recomposition. The path mutation (push / pop)
    // is deferred to LaunchedEffect since mutating SnapshotState during
    // composition is forbidden.
    if (currentUri.isNotEmpty()) {
        coordinator.cache(currentUri, node)
    }
    LaunchedEffect(node) {
        if (currentUri.isNotEmpty()) {
            coordinator.receive(currentUri, node)
        }
    }

    val rootUri by coordinator.rootUri
    val path: SnapshotStateList<String> = coordinator.path

    // Effective top-of-stack URI. Falls back to currentUri on the very
    // first publish before the coordinator has seeded its root.
    val activeUri = path.lastOrNull() ?: rootUri ?: currentUri
    val isRoot = path.isEmpty()

    // Resolve the active level's content via the cache; fall back to
    // the live `node` only when the cache hasn't been populated yet
    // (initial publish, before LaunchedEffect's receive ran).
    val activeNode = coordinator.rootNodeCache[activeUri] ?: node

    val title = activeNode.props.getString("title", "")
    val subtitle = activeNode.props.getString("subtitle", "")
    val showBack = activeNode.props.getBool("back")
    val bgArgb = activeNode.props.getColor("background_color", 0)
    val textArgb = activeNode.props.getColor("text_color", 0)
    val hasExplicitBg = bgArgb != 0
    val hasExplicitText = textArgb != 0

    // Filter children for actions and the screen content body.
    val actions = activeNode.children.filter { it.type == "top_bar_action" }
    val screenContent = activeNode.children.firstOrNull {
        it.type != "top_bar_action" && !NativeRootHostRegistry.consumes(it.type)
    }

    // Inline search field config (NavBar::searchBar() — Apple HIG /
    // Expo pattern). When set, replaces the title slot with a search
    // text field; query changes flow back via TEXT_CHANGE.
    val searchPlaceholder = activeNode.props.getString("search_placeholder", "")
    val searchOnQueryCb = activeNode.props.getCallbackId("search_on_query")
    val searchDebounceMs = activeNode.props.getInt("search_debounce_ms", 300)
    val hasSearch = searchPlaceholder.isNotEmpty()

    // System back / predictive-back: shrink path if pushed, otherwise
    // forward to PHP so it pops the underlying stack (e.g. back to
    // wherever the user came from before entering native chrome).
    BackHandler(enabled = true) {
        if (path.isNotEmpty()) {
            path.removeLast()
            coordinator.onPathChange(path.toList())
        } else {
            NativeElementBridge.sendSystemBackEvent()
        }
    }

    Scaffold(
        topBar = {
            TopAppBar(
                title = {
                    if (hasSearch) {
                        InlineNavSearchField(
                            placeholder = searchPlaceholder,
                            callbackId = searchOnQueryCb,
                            nodeId = activeNode.id,
                            debounceMs = searchDebounceMs,
                        )
                    } else if (subtitle.isNotEmpty()) {
                        Column {
                            Text(
                                title,
                                style = MaterialTheme.typography.titleMedium,
                                fontWeight = FontWeight.SemiBold
                            )
                            Text(
                                subtitle,
                                style = MaterialTheme.typography.labelSmall
                            )
                        }
                    } else {
                        Text(title, fontWeight = FontWeight.SemiBold)
                    }
                },
                navigationIcon = {
                    if (showBack) {
                        IconButton(onClick = {
                            // Manual back chevron always defers to system back —
                            // shrinks path if pushed, otherwise tells PHP.
                            if (path.isNotEmpty()) {
                                path.removeLast()
                                coordinator.onPathChange(path.toList())
                            } else {
                                NativeElementBridge.sendSystemBackEvent()
                            }
                        }) {
                            Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = "Back")
                        }
                    }
                },
                actions = {
                    actions.forEach { action ->
                        TopBarActionView(action)
                    }
                },
                colors = if (hasExplicitBg) {
                    val bg = argbToComposeColor(bgArgb)
                    val fg = if (hasExplicitText) argbToComposeColor(textArgb) else Color.White
                    TopAppBarDefaults.topAppBarColors(
                        containerColor = bg,
                        titleContentColor = fg,
                        navigationIconContentColor = fg,
                        actionIconContentColor = fg
                    )
                } else {
                    TopAppBarDefaults.topAppBarColors()
                }
            )
        },
        modifier = modifier.fillMaxSize()
    ) { padding ->
        // AnimatedContent slides between stack levels. Direction
        // discrimination (push vs pop) would need state we don't have
        // handy here; the symmetric slide is acceptable for now.
        AnimatedContent(
            targetState = activeUri,
            transitionSpec = {
                val intSpec = tween<androidx.compose.ui.unit.IntOffset>(durationMillis = 250)
                slideInHorizontally(intSpec) { it } togetherWith
                    slideOutHorizontally(intSpec) { -it }
            },
            label = "stack-level",
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
        ) { uri ->
            val levelNode = coordinator.rootNodeCache[uri]
            val levelContent = levelNode?.children?.firstOrNull {
                it.type != "top_bar_action" && !NativeRootHostRegistry.consumes(it.type)
            }
            if (levelContent != null) {
                NodeView(node = levelContent)
            } else if (uri == currentUri && screenContent != null) {
                NodeView(node = screenContent)
            } else {
                Box(modifier = Modifier.fillMaxSize())
            }
        }
    }
}

/**
 * Renders a single trailing toolbar action — plain IconButton when no
 * sub-items, DropdownMenu when `NavAction::items()` produced
 * `top_bar_action` children. Module-internal so the tabs renderer can
 * reuse the same plumbing when the layout folds NavBar actions onto a
 * `native_root_tabs` sentinel.
 */
@Composable
internal fun TopBarActionView(action: NativeUINode) {
    val icon = action.props.getString("icon", "more_vert")
    val subItems = action.children.filter { it.type == "top_bar_action" }

    if (subItems.isEmpty()) {
        IconButton(onClick = {
            if (action.onPress != 0) {
                NativeElementBridge.sendPressEvent(action.onPress, action.id)
            }
        }) {
            MaterialIcon(name = icon, contentDescription = action.props.getString("label", ""))
        }
    } else {
        var expanded by remember { mutableStateOf(false) }
        IconButton(onClick = { expanded = true }) {
            MaterialIcon(name = icon, contentDescription = action.props.getString("label", ""))
        }
        DropdownMenu(expanded = expanded, onDismissRequest = { expanded = false }) {
            subItems.forEach { item ->
                if (item.props.getBool("divider")) {
                    // Inline visual separator emitted by `NavAction::divider()`.
                    HorizontalDivider()
                    return@forEach
                }
                val itemLabel = item.props.getString("label", "")
                val itemIcon = item.props.getString("icon", "")
                val isDestructive = item.props.getBool("destructive")
                val destructiveColor = MaterialTheme.colorScheme.error
                DropdownMenuItem(
                    text = {
                        Text(
                            itemLabel,
                            color = if (isDestructive) destructiveColor else Color.Unspecified
                        )
                    },
                    leadingIcon = if (itemIcon.isNotEmpty()) {
                        {
                            // `destructive()` should tint the whole row red — text +
                            // icon — to mirror SwiftUI's `Button(role: .destructive)`
                            // which colors both. Using the bare default leaves the
                            // icon at LocalContentColor while the text reads red,
                            // which scans as a styling glitch.
                            if (isDestructive) {
                                MaterialIcon(
                                    name = itemIcon,
                                    contentDescription = itemLabel,
                                    tint = destructiveColor,
                                )
                            } else {
                                MaterialIcon(name = itemIcon, contentDescription = itemLabel)
                            }
                        }
                    } else null,
                    onClick = {
                        expanded = false
                        if (item.onPress != 0) {
                            NativeElementBridge.sendPressEvent(item.onPress, item.id)
                        }
                    }
                )
            }
        }
    }
}
